xdgopenproxy/portal_launcher.go

```diff
@@ -0,0 +1,174 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package xdgopenproxy
+
+import (
+	"fmt"
+	"syscall"
+	"time"
+
+	"github.com/godbus/dbus"
+)
+
+const (
+	desktopPortalBusName      = "org.freedesktop.portal.Desktop"
+	desktopPortalObjectPath   = "/org/freedesktop/portal/desktop"
+	desktopPortalOpenURIIface = "org.freedesktop.portal.OpenURI"
+	desktopPortalRequestIface = "org.freedesktop.portal.Request"
+)
+
+// portalLauncher is a launcher that forwards the requests to xdg-desktop-portal DBus API
+type portalLauncher struct{}
+
+// desktopPortal gets a reference to the xdg-desktop-portal D-Bus service
+func (p *portalLauncher) desktopPortal(bus *dbus.Conn) (dbus.BusObject, error) {
+	// We call StartServiceByName since old versions of
+	// xdg-desktop-portal do not include the AssumedAppArmorLabel
+	// key in their service activation file.
+	var startResult uint32
+	err := bus.BusObject().Call("org.freedesktop.DBus.StartServiceByName", 0, desktopPortalBusName, uint32(0)).Store(&startResult)
+	if dbusErr, ok := err.(dbus.Error); ok {
+		// If it is not possible to activate the service
+		// (i.e. there is no .service file or the systemd unit
+		// has been masked), assume it is already
+		// running. Subsequent method calls will fail if this
+		// assumption is false.
+		if dbusErr.Name == "org.freedesktop.DBus.Error.ServiceUnknown" || dbusErr.Name == "org.freedesktop.systemd1.Masked" {
+			err = nil
+			startResult = 2
```

Comment:
Maybe a comment here like "// DBUS_START_REPLY_ALREADY_RUNNING" or using constant ? But not a blocker as this is all close together.